crates/circuit/src/circuit_data.rs

```diff
@@ -396,6 +397,46 @@ impl CircuitData {
         }
         Ok(())
     }
+
+    /// Get an immutable view of the instructions in the circuit data
+    pub fn data(&self) -> &[PackedInstruction] {
+        &self.data
+    }
+
+    /// Clone an empty CircuitData from a given reference.
+    ///
+    /// The new copy will have the global properties from the provided `CircuitData`.
+    /// The the bit data fields and interners, global phase, etc will be copied to
+    /// the new returned `CircuitData`, but the `data` field's instruction list will
+    /// be empty. This can be useful for scenarios where you want to rebuild a copy
+    /// of the circuit from a reference but insert new gates in the middle.
+    pub fn clone_empty_from(other: &Self) -> Self {
+        CircuitData {
+            data: Vec::with_capacity(other.data.len()),
```

Comment:
This looks like a potential sharp bit, as we don't really know how large the new circuit is going to be. How about making the capacity and option than defaults to the other circuit's size?

Reply:
I made this configurable and explicit in: https://github.com/Qiskit/qiskit/pull/13331/commits/34c3e918e291d2a039289fc2c0d69c88721763f5